Dynamo - config gated torch.distributed allow, exclusion for special leaf funcs

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110894

`is_allowed` is a tricky bit of functionality - it sits early up in builder and is used to drive the creation of TorchVariable (more notes here, meta only https://fb.workplace.com/groups/pytorch.dev/permalink/1393563781222098/) 

If we are tracing distributed in full, we want to route certain calls in distributed to NOT PASS is_allowed (this does not, confusingly, mean that they are not allowed, lol, but rather that we dont want them to become TorchVariable), others, we are fine with preserving.



cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

voznesenskym committed:
distributed skipfile and allow check

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/skipfiles.py:
```diff
@@ -196,6 +196,11 @@ def _module_dir(m: types.ModuleType):
     _module_dir(torch) + "_export/wrappers.py",
 }
 
+if torch.distributed.is_available():
```
config this too, I guess.

voznesenskym replied:
Actually, I am not sure this matters, let's see if tests fail, the inline vs allow refactor has made this a little more confusing.

voznesenskym committed:
Update on "Dynamo - config gated is_allowed routing, skipfiles for distributed"


`is_allowed` is a tricky bit of functionality - it sits early up in builder and is used to drive the creation of TorchVariable (more notes here, meta only https://fb.workplace.com/groups/pytorch.dev/permalink/1393563781222098/) 

If we are tracing distributed in full, we want to route certain calls in distributed to NOT PASS is_allowed (this does not, confusingly, mean that they are not allowed, lol, but rather that we dont want them to become TorchVariable), others, we are fine with preserving.



cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo - config gated is_allowed routing, skipfiles for distributed"


`is_allowed` is a tricky bit of functionality - it sits early up in builder and is used to drive the creation of TorchVariable (more notes here, meta only https://fb.workplace.com/groups/pytorch.dev/permalink/1393563781222098/) 

If we are tracing distributed in full, we want to route certain calls in distributed to NOT PASS is_allowed (this does not, confusingly, mean that they are not allowed, lol, but rather that we dont want them to become TorchVariable), others, we are fine with preserving.



cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
I'm signing off in terms of functional correctness for the hack.

However, @yanboliang should have the final say here, since this increases his workload for the refactor.

yanboliang commented on torch/_dynamo/skipfiles.py:
```diff
@@ -196,6 +196,11 @@ def _module_dir(m: types.ModuleType):
+    FILENAME_INLINELIST |= set(
+        glob.glob(_module_dir(torch) + "distributed/**/*.py", recursive=True),
+    )
+
```
Can you add ```torch.distributed``` into the ```SUBMODULE_INLINELIST```? That would be a more easy way to force inline all files under a submodule.

voznesenskym replied:
No, this fails for whatever reason.

voznesenskym replied:
If you want a crack at debugging:

1) voz/fsdp_autograd3
2) build
3) torchrun --standalone --nproc_per_node=2 fsdp.py

yanboliang replied:
I think this would break trunk since you land after #110835. Do you mind to send a forward fix? Or I can help to forward fix it.

yanboliang commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -307,7 +307,40 @@ def _builtin_constant_ids():
 
+# A subcheck of is_allowed, we utilize this for patching is_allowed around distributed.
+# We do this because we want to allow these to be traced, and hence covered in skipfiles, but we do not want them to
+# become TorchVariable
```
If you don't want to make them as ```TorchVariable```, we should not make them ```is_allowed``` returning True after my refactor. If you just want to inline these functions, can you treat them as regular python functions and go through the regular inline rules?

voznesenskym replied:
We need some to go one way, some the other, the logic here is correct - and they shouldn't go through inline rules.

voznesenskym replied:
The problem is they are under torch.* so they get routed to TorchVariable - for some like

```
 if obj in [
            torch.distributed._functional_collectives_impl._all_gather_into_tensor,
            torch.distributed._functional_collectives_impl._all_reduce,
            torch.distributed._functional_collectives_impl._reduce_scatter_tensor,
            torch.distributed._functional_collectives_impl._all_reduce_coalesced,
            torch.distributed._functional_collectives_impl._all_gather_into_tensor_coalesced,
            torch.distributed._functional_collectives_impl._reduce_scatter_tensor_coalesced,
        ]:
```
It is *correct* for them to become TorchVariable

For others (the rest of the checks) - we need to make sure they *do not* become TorchVariable. If we don't have this function, the wrong types will become TorchVariable instead of passing through `CollectiveFunctionRewriteVariable.can_rewrite(value)` `CollectiveFunctionRewriteVariable.can_rewrite(value)` etc

ezyang replied:
I don't quite understand Voz's explanation, but I certainly agree with Voz that there is something funny going on here.

To give an alternate example, on main, I'd like to inline into functions in `_functional_collectives`. I applied this patch:

```
diff --git a/torch/_dynamo/allowed_functions.py b/torch/_dynamo/allowed_functions.py
index 8beca2b4502..d155023ef54 100644
--- a/torch/_dynamo/allowed_functions.py
+++ b/torch/_dynamo/allowed_functions.py
@@ -175,6 +175,7 @@ def _allowed_function_ids():
             # issues observed in
             # https://github.com/pytorch/pytorch/issues/108269
             "torch.distributed.algorithms.",
+            "torch.distributed._functional_collectives.",
         )
         allowed_modules_dot = tuple([x + "." for x in allowed_modules])
         module = inspect.getmodule(obj)
```

But it does not work; for some reason I appear to still be trying to place things like `all_to_all_single` directly into the graph. With Voz's PR and my patch, and turning on `trace_distributed`, only then do I get the expected behavior of inlining.

wconstab replied:
more of a meta question but do we have a design doc for "the new skipfiles/allowed_functions design"? I am a little wary of all this complexity around distributed, hopefully we can do it more cleanly in a redesign

voznesenskym committed:
Update on "Dynamo - config gated is_allowed routing, skipfiles for distributed"


`is_allowed` is a tricky bit of functionality - it sits early up in builder and is used to drive the creation of TorchVariable (more notes here, meta only https://fb.workplace.com/groups/pytorch.dev/permalink/1393563781222098/) 

If we are tracing distributed in full, we want to route certain calls in distributed to NOT PASS is_allowed (this does not, confusingly, mean that they are not allowed, lol, but rather that we dont want them to become TorchVariable), others, we are fine with preserving.



cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo - config gated is_allowed routing, skipfiles for distributed"


`is_allowed` is a tricky bit of functionality - it sits early up in builder and is used to drive the creation of TorchVariable (more notes here, meta only https://fb.workplace.com/groups/pytorch.dev/permalink/1393563781222098/) 

If we are tracing distributed in full, we want to route certain calls in distributed to NOT PASS is_allowed (this does not, confusingly, mean that they are not allowed, lol, but rather that we dont want them to become TorchVariable), others, we are fine with preserving.



cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -346,6 +349,8 @@ def is_fbcode():
     [name for name, _ in inspect.getmembers(torch.Tensor) if re.match(r"^is_.*", name)]
 )
 
+trace_distributed = True
```
mistake

ezyang commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -233,6 +236,17 @@ def _find_torch_objects(module):
     _find_torch_objects(torch)
     _find_torch_objects(math)
 
+    if config.trace_distributed:
```
do you have to import torch.distribute._fun... here?

voznesenskym replied:
It seems like we do not.

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@yanboliang, revert if its not up to snuff, nbd nbd :)

voznesenskym commented:
@pytorchbot merge
